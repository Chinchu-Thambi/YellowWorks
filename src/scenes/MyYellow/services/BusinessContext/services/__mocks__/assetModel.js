const assetModel = {
  caption: 'caption',
  contentUrl: 'contentUrl',
  encodingFormat: 'encodingFormat',
  height: 'height',
  thumbnailUrl: 'thumbnailUrl',
  type: 'type',
  width: 'width',
};

export default assetModel;
