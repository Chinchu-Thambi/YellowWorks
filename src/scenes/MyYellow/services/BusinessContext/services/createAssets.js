import axios from 'axios';
import * as R from 'ramda';

const endpoint = process.env.GATSBY_PRODUCT_API_ENDPOINT ?? process.env.STORYBOOK_PRODUCT_API_ENDPOINT;

const query = `
mutation CreateAsset(
  $caption: String!,
  $contentUrl: AWSURL!,
  $businessId: ID!,
  $customerId: ID
) {
  createAsset(
    businessId: $businessId, 
    customerId: $customerId, 
    input: {
      caption: $caption,
      contentUrl: $contentUrl,
      type: ImageObject
    }
  ) {
    caption
    contentUrl
    encodingFormat
    height
    thumbnailUrl
    type
    width
  }
}
`;

/** @type {(
 * jwtToken: string, customerId: string, businessId: string
 * ) => (
 * files: {url: string}[]
 * ) => ({
 *  caption: string
 *  contentUrl: string
 *  encodingFormat: string
 *  height: string
 *  thumbnailUrl: string
 *  type: string
 *  width: string
 * }[])} */
const createAssets = (jwtToken, customerId, businessId) => (files) => {
  const variables = {
    businessId,
    customerId,
  };

  const axiosPost = (file) => axios.post(
    endpoint,
    {
      query,
      variables: {
        ...variables,
        contentUrl: file.url,
        caption: file.file.name,
      },
    },
    {
      headers: {
        Authorization: jwtToken,
      },
    },
  );

  return Promise.all(
    R.map(axiosPost)(files),
  ).then(
    R.map(R.path(['data', 'data', 'createAsset'])),
  );
};

export default createAssets;
