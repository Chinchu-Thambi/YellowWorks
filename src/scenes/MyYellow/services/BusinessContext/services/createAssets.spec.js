
import axios from 'axios';

import createAssets from './createAssets';

import assetModel from './__mocks__/assetModel';

describe('createAssets', () => {
  const token = 'token';
  const customerId = 'customerId';
  const businessId = 'businessId';

  const files = [
    { file: new File([''], 'file name'), url: 'file url', id: 'file id' },
    { file: new File([''], 'file name 2'), url: 'file url 2', id: 'file id 2' },
  ];

  axios.post.mockImplementation(() => ({
    data: {
      data: {
        createAsset: assetModel,
      },
    },
  }));

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('given a token, customer id and business id,  returns a function', () => {
    expect(typeof createAssets(token, customerId, businessId)).toBe('function');
  });

  it('posts an axios graphql query', async () => {
    await createAssets(token, customerId, businessId)(files);

    expect(axios.post).toBeCalledTimes(files.length);
    expect(axios.post.mock.calls[0][1].query).toContain('mutation CreateAsset');
    expect(axios.post.mock.calls[0][1].query).toMatchSnapshot();
    expect(axios.post.mock.calls[0][1].variables).toEqual({
      caption: files[0].file.name, businessId, customerId, contentUrl: files[0].url,
    });
    expect(axios.post.mock.calls[0][2].headers.Authorization).toBe(token);

    expect(axios.post.mock.calls[1][1].variables).toEqual({
      caption: files[1].file.name, businessId, customerId, contentUrl: files[1].url,
    });
  });

  it('transforms axios post in createdAsset', async () => {
    const createdAssets = await createAssets(token, customerId, businessId)(files);

    expect(createdAssets[0]).toEqual(assetModel);
  });
});
