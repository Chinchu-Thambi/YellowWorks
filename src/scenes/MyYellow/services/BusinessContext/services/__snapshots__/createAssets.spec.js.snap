// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAssets posts an axios graphql query 1`] = `
"
mutation CreateAsset(
  $caption: String!,
  $contentUrl: AWSURL!,
  $businessId: ID!,
  $customerId: ID
) {
  createAsset(
    businessId: $businessId, 
    customerId: $customerId, 
    input: {
      caption: $caption,
      contentUrl: $contentUrl,
      type: ImageObject
    }
  ) {
    caption
    contentUrl
    encodingFormat
    height
    thumbnailUrl
    type
    width
  }
}
"
`;
